File: blobExec/src/main/scala/cregit/blobexec/TokenizerIdentity.scala
package cregit.blobexec

/** Which tokenizer produced the tokens for each file extension.
  *
  * The problem this type exists for. `blob_map` reuse was decided on `command`
  * and `mask` alone (see `Mapping.open`), and `command` is the fixed path
  * `tokenizeByBlobId/tokenBySha.pl` — a constant for the life of the pipeline.
  * So when the Rust tokenizer's output format was corrected in 729643e, nothing
  * about the recorded metadata changed, every cached `.rs` row stayed a cache
  * hit, and a run with the corrected tokenizer reproduced the old, wrong tokens
  * exactly. The mask cannot stand in for this: the mask decides WHICH files are
  * tokenized and never HOW.
  *
  * The identity is per EXTENSION rather than per language, because the extension
  * is what the cache rows can be selected by: `blob_map` stores paths, not
  * languages. `CregitLanguages.pm` maps extension to language and language to
  * parser, so the pipeline collapses that chain into one value per extension
  * before handing it over (see `tokenize/tokenizerIdentity.pl`).
  *
  * The value is opaque here on purpose. This code never needs to know whether it
  * is a hash of a binary, a version string the tokenizer reports, or a digest of
  * a whole toolchain — only whether it is the same string as last time. */
final case class TokenizerIdentity(byExtension: Map[String, String]) {

  def isEmpty: Boolean = byExtension.isEmpty
  def nonEmpty: Boolean = byExtension.nonEmpty
  def extensions: Set[String] = byExtension.keySet
  def get(extension: String): Option[String] = byExtension.get(extension)

  /** Canonical, sorted rendering: the same map always prints the same string, so
    * it can go in a log or a refusal message and be compared by eye. */
  def render: String =
    byExtension.toVector.sorted.map { case (e, id) => s"$e=$id" }.mkString(",")
}

object TokenizerIdentity {

  val empty: TokenizerIdentity = TokenizerIdentity(Map.empty)

  /** Extensions as `CregitLanguages.pm` spells them: lowercase, no leading dot.
    * `c++` and `h++` are real entries in that table, hence the `+`. The pattern
    * is also a safety property: these strings are interpolated into a SQL LIKE
    * pattern in [[Mapping]], and this alphabet contains no quote, no `%` and no
    * `_`, so there is nothing to escape and nothing to inject. */
  private[blobexec] val ExtensionPattern = "[a-z0-9+]+"

  /** The identity value: hex, and long enough to be a digest rather than a
    * guess. Bounded above so a malformed argument cannot become a meta row of
    * arbitrary size. */
  private[blobexec] val ValuePattern = "[0-9a-f]{8,128}"

  private val EntryRe = s"($ExtensionPattern)=($ValuePattern)".r

  /** Parse `ext=value[,ext=value]...`.
    *
    * Left on anything malformed, naming the offending entry. Refusing a
    * malformed identity matters more than it looks: a value this code does not
    * understand would be recorded verbatim, compare unequal to the next run's,
    * and turn every subsequent run into a refusal nobody can explain. */
  def parse(spec: String): Either[String, TokenizerIdentity] = {
    val trimmed = spec.trim
    if (trimmed.isEmpty)
      return Left("a tokenizer identity must name at least one extension, as ext=value")
    val entries = trimmed.split(",", -1).toVector
    val parsed = entries.map {
      case EntryRe(ext, value) => Right((ext, value))
      case other =>
        Left(
          s"cannot read [$other] as a tokenizer identity entry. Wanted " +
            s"<extension>=<value>, extension matching $ExtensionPattern (lowercase, no " +
            s"leading dot, as in CregitLanguages.pm) and value matching $ValuePattern."
        )
    }
    parsed.collectFirst { case Left(why) => Left(why) }.getOrElse {
      val pairs = parsed.collect { case Right(p) => p }
      val duplicated = pairs.groupBy(_._1).collect { case (e, vs) if vs.size > 1 => e }
      if (duplicated.nonEmpty)
        Left(s"extension(s) named more than once in the tokenizer identity: " +
          duplicated.toVector.sorted.mkString(", "))
      else Right(TokenizerIdentity(pairs.toMap))
    }
  }

  /** Parse a `--retokenize=` extension list. Same alphabet as above, so the two
    * cannot disagree about what an extension looks like. */
  def parseExtensions(spec: String): Either[String, Set[String]] = {
    val trimmed = spec.trim
    if (trimmed.isEmpty)
      return Left("--retokenize must name at least one extension, as in --retokenize=rs")
    val parts = trimmed.split(",", -1).toVector.map(_.trim)
    val bad = parts.filterNot(_.matches(ExtensionPattern))
    if (bad.nonEmpty)
      Left(s"not an extension: ${bad.map(b => s"[$b]").mkString(", ")}. Wanted lowercase " +
        s"names without a leading dot, matching $ExtensionPattern (as in CregitLanguages.pm): " +
        "--retokenize=rs, --retokenize=c,h")
    else Right(parts.toSet)
  }
}

File: tokenize/tokenizerIdentity.pl
#!/usr/bin/env perl

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Prints blobExec's --tokenizer-identity argument: one opaque value per file
# extension, changing whenever the thing that turns that extension's bytes into
# tokens changes.
#
#   ext=<hex>,ext=<hex>,...
#
# Why this exists. blobExec decided whether to reuse a cached tokenization on
# `command` and `mask`. `command` is the fixed path
# tokenizeByBlobId/tokenBySha.pl and `mask` says WHICH files to tokenize, never
# HOW. So when the Rust tokenizer's output format was corrected in 729643e,
# neither value moved: every cached .rs row stayed a cache hit and a run with
# the corrected binary reproduced the broken tokens, in 45 projects, with no
# error anywhere. An identity that tracks the TOOLCHAIN is the missing key.
#
# What goes into an extension's identity, and why each part:
#
#   * its language's parser (CregitLanguages::LANG_PARSER_REL) — the tokenizer
#     itself. For Rust that is the compiled cargo artifact, which is exactly the
#     file that went stale.
#   * srcml2token, srcml and ctags, for the srcML-routed languages only. The
#     token stream is the product of all three: srcml2token transcodes srcML's
#     XML, and tokenizeSrcMl.pl merges ctags output into it. A change in any of
#     them changes the tokens.
#   * tokenize.pl and CregitLanguages.pm, for every extension. tokenize.pl is
#     the dispatcher that decides which parser runs and with which flags
#     (--position among them, which the token format depends on), and
#     CregitLanguages.pm decides which language an extension is. Either one can
#     change an extension's tokens without the parser changing at all.
#
# Including the shared files makes the identity CONSERVATIVE: editing
# tokenize.pl moves every extension's identity, so every project with cached
# rows refuses until an operator decides what to do. That is the right direction
# to be blunt in. A refusal costs a conversation; a false "unchanged" costs a
# republished corpus.
#
# The output is sorted and every value is a sha256 hex digest of a digest list,
# so the string is byte-stable for a given checkout: it is stored in the blob
# map's meta table and compared string-for-string on every resume.

use strict;
use warnings;

use Digest::SHA qw(sha256_hex);
use File::Basename;
use FindBin;
use lib $FindBin::Bin;
use CregitLanguages;
use Getopt::Long;

my $usage = "
Usage: $0 [options]

Prints blobExec's --tokenizer-identity value for this checkout:
  ext=<sha256>,ext=<sha256>,...

Options:
   --srcml2token=<path>   the built C++ transcoder (required for C/C++/Java)
   --srcml=<path>         the srcml binary       (required for C/C++/Java)
   --ctags=<path>         the ctags binary       (required for C/C++/Java)
   --all-extensions       cover every extension in the table, not only the
                          masked ones (.am/.ac are routed but not masked)
";

my $srcml2tokenPath = "";
my $srcmlPath       = "";
my $ctagsPath       = "";
my $allExtensions   = 0;

GetOptions(
    "srcml2token=s"  => \$srcml2tokenPath,
    "srcml=s"        => \$srcmlPath,
    "ctags=s"        => \$ctagsPath,
    "all-extensions" => \$allExtensions,
) or die($usage);

my $basedir = $FindBin::Bin;

# Every extension's identity includes these, because either one can change what
# an extension's tokens are without any parser changing.
my @shared = ("$basedir/tokenize.pl", "$basedir/CregitLanguages.pm");

my %parsers = CregitLanguages::parsers($basedir);

my @extensions = $allExtensions
    ? CregitLanguages::extensions_sorted()
    : CregitLanguages::masked_extensions_sorted();

# A file that is named but absent is fatal. Emitting an identity that silently
# omitted a component would be worse than useless: it would compare EQUAL across
# a change in that component, which is the exact failure this program exists to
# make impossible. The Rust parser is a build artifact, so this is also where an
# unbuilt checkout is caught.
sub digest_of {
    my ($path) = @_;
    open(my $fh, '<', $path)
        or die "tokenizerIdentity: cannot read [$path]: $!\n"
             . "  It is part of a tokenizer's identity, and an identity computed without it\n"
             . "  would compare equal across a change in it. Build the checkout first\n"
             . "  (./run_pipeline_process.sh --ensure-artifacts) and try again.\n";
    binmode $fh;
    my $sha = Digest::SHA->new(256);
    $sha->addfile($fh);
    close $fh;
    return $sha->hexdigest;
}

my %digestCache;
sub cached_digest {
    my ($path) = @_;
    $digestCache{$path} //= digest_of($path);
    return $digestCache{$path};
}

my @pairs;
for my $ext (@extensions) {
    my $language = $CregitLanguages::EXT_LANG{$ext};
    die "tokenizerIdentity: no language for extension [$ext]\n" unless defined $language;
    my $parser = $parsers{$language};
    die "tokenizerIdentity: no parser for language [$language]\n" unless defined $parser;

    my @components = (@shared, $parser);

    # The srcML chain is three binaries deep and all three shape the tokens.
    # Required rather than optional: an identity for .c computed without
    # srcml2token would not move when srcml2token does.
    if ($CregitLanguages::LANG_PARSER_REL{$language} eq $CregitLanguages::SRCML_PARSER_REL) {
        for my $pair (["--srcml2token", $srcml2tokenPath],
                      ["--srcml",       $srcmlPath],
                      ["--ctags",       $ctagsPath]) {
            my ($flag, $path) = @$pair;
            die "tokenizerIdentity: $flag is required, because extension [$ext] is parsed by\n"
              . "  $CregitLanguages::SRCML_PARSER_REL and its tokens depend on that binary.\n$usage"
                if $path eq "";
            push @components, $path;
        }
    }

    # Digest of the component digests, not of the concatenated bytes: the list is
    # what identifies the toolchain, and a per-file digest makes a failure
    # readable with --verbose-style debugging if it is ever needed.
    my $combined = sha256_hex(join("\n", map { cached_digest($_) } @components));
    push @pairs, "$ext=$combined";
}

die "tokenizerIdentity: no extensions to report\n" unless @pairs;

print join(",", @pairs), "\n";
